port/posix/posix_transport_shm.c

```diff
@@ -1,301 +1,581 @@
-    posixTransportShmConfig*  config = (posixTransportShmConfig*)cf;
-
-    if (ctx == NULL || config == NULL || config->shmObjName == NULL) {
+    int ret = WH_ERROR_OK;
+    void* ptr = NULL;
+
+    if (    (fd < 0) ||
+            (map == NULL) ) {
         return WH_ERROR_BADARGS;
     }
 
-    /* check shmObjName for max length */
-    shmNameLen = strnlen(config->shmObjName, NAME_MAX + 1);
-    if (shmNameLen > NAME_MAX) {
-        return WH_ERROR_BADARGS;
+    /* Map the shared memory object */
+    ptr = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+    if (ptr != MAP_FAILED) {
+        memset(map, 0, sizeof(*map));
+        map->ptr = ptr;
+        map->size = size;
+        map->header = (ptshmHeader*)ptr;
+        map->req = (uint8_t*)(map->header + 1);
+        map->resp = map->req + map->header->req_size;
+        map->dma = map->resp + map->header->resp_size;
+    } else {
+        ret = WH_ERROR_ABORTED;
     }
+    return ret;
+}
 
-    ctx->shmObjName = calloc(shmNameLen + 1, sizeof(char));
-    strncpy(ctx->shmObjName, config->shmObjName, shmNameLen);
 
-    shmSize = sizeof(ShmHeader) + config->req_size + config->resp_size;
+static int posixTransportShm_UseMap(char* name, ptshmMapping* map)
+{
+    int ret = WH_ERROR_OK;
+    int fd = -1;
 
-    /* Attempt to open the shared memory object serving as the backing store
-     * without attempting to create it. If it already exists, attempt to delete
-     * it, and verify that it is gone. */
-    shmFd = shm_open(ctx->shmObjName, O_RDWR, 0666);
-    if (shmFd > 0) {
-        close(shmFd);
-        shm_unlink(ctx->shmObjName);
-        shmFd = shm_open(ctx->shmObjName, O_RDWR, 0666);
+    if( (name == NULL) || (map == NULL)) {
+        return WH_ERROR_BADARGS;
     }
 
-    /* If shared memory object doesn't exist, create it and set the size */
-    if (shmFd == -1) {
-        if (errno == ENOENT) {
-            /* Shared memory object doesn't exist, create it */
-            shmFd = shm_open(ctx->shmObjName, O_CREAT | O_RDWR, 0666);
-            if (shmFd == -1) {
-                perror("server shm_open");
-                exit(EXIT_FAILURE);
-            }
-
-            /* Set the size of the shared memory object */
-            if (ftruncate(shmFd, shmSize) == -1) {
-                perror("ftruncate");
-                exit(EXIT_FAILURE);
+    fd = shm_open(name, O_RDWR, 0);
+    if (fd >= 0) {
+        /* Check the size */
+        struct stat st[1] = { 0 };
+        ret = fstat(fd, st);
+        if (ret == 0) {
+            if (st->st_size != 0) {
+                /* Map the header and get configuration */
+                ptshmHeader* header = (ptshmHeader*)mmap(NULL, sizeof(*header),
+                        PROT_READ, MAP_SHARED, fd, 0);
+                if (header != MAP_FAILED) {
+                    size_t size = 0;
+                    if(header->initialized > PTSHM_INITIALIZED_NONE) {
```

Comment:
perhaps an explicit check for `if (header->initialized == PTSHM_INITIALIZED_CREATOR)`, since I think all other states should be an invalid sequence, right? 

Also, there is a missing space after `if`

Reply:
corrected